crypto/threads_pthread.c

```diff
@@ -655,6 +820,15 @@ int CRYPTO_THREAD_unlock(CRYPTO_RWLOCK *lock)
 # ifdef USE_RWLOCK
     if (pthread_rwlock_unlock(lock) != 0)
         return 0;
+#  ifdef REPORT_RWLOCK_CONTENTION
+    {
+        struct stack_traces *traces = CRYPTO_THREAD_get_local(&thread_contention_data);
+
+        if (contention_fp == NULL)
+            return 1;
+        print_stack_traces(traces, contention_fp);
```

Comment:
It's kind of strange that we defer printing stack traces until we unlock, but then print *all* the stack traces in one go on first unlock. My assumption is that we are avoiding printing the stack traces until unlock to avoid deadlock scenarios (i.e. avoid taking the contention lock mutex while holding another lock), and to reduce the amount of time we are holding locks for.

So, lets say we are in some stack trace and holding onto three read locks at the same time. As we take each one we add it to the list of stack traces and increment `traces->idx`. Then when we finish the first unlock (i.e. we are still holding two of the locks) we then print all the stack traces. This seems really odd. If we are deferring until unlock, I would have expected to defer until *all* the locks are unlocked.